packages/messaging/src/helpers/externalizePayload.ts

```diff
@@ -0,0 +1,98 @@
+/* eslint-disable camelcase */
+/**
+ * @license
+ * Copyright 2017 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+  MessagePayload,
+  MessagePayloadInternal
+} from '../interfaces/message-payload';
+
+export function externalizePayload(
+  internalPayload: MessagePayloadInternal
+): MessagePayload {
+  const payload: MessagePayload = {
+    notification: {},
+    data: {},
+    fcmOptions: {},
+    from: internalPayload.from,
+    collapseKey: internalPayload.collapse_key
+  } as MessagePayload;
+
+  propagateNotificationPayload(payload, internalPayload);
+  propagateDataPayload(payload, internalPayload);
+  propagateFcmOptions(payload, internalPayload);
+
+  return payload;
+}
+
+function propagateNotificationPayload(
+  payload: MessagePayload,
+  messagePayloadInternal: MessagePayloadInternal
+): void {
+  if (!messagePayloadInternal.notification) {
+    delete payload.notification;
```

Comment:
Instead of defining an empty object for `payload.notification`, then deleting it, it would be more efficient(marginally though) to only create an object and assign it to `payload.notification` if `messagePayloadInternal.notification` exists.

The same applies to `payload.data` and `payload.fcmOptions` as well.